Add gh-pages with JSDocs and demo app

lilaconlee opened the pull request:
### Summary

Add JSDocs and a demo app to the gh-pages. Moved significant chunks of the README to the JSDoc landing page. This doesn't include a publishing script, I figure we'll do it manually the first time and then I'll PR a lil script. 

Building gh-pages generates the following in `dist`:
```
dist/gh-pages/
├── assign-polyfill.js
├── docs
│   ├── 1.0.2
│   ├── current -> 1.0.2
│   └── index.html - redirect to /current
└── index.html - demo app
```

crookedneighbor commented on jsdoc/Home.md:
````diff
@@ -0,0 +1,90 @@
+You can also get Drop-in from [npm](https://www.npmjs.com/).
+
+```sh
+npm install --save braintree-web-drop-in
````
This should probably specify:

```
var dropin = require('braintree-web-drop-in');

dropin.create({ /* options //* }, callback);
```

lilaconlee replied:
Oh I like doing this and the previous comment actually.

crookedneighbor commented on jsdoc/Home.md:
````diff
@@ -0,0 +1,90 @@
+The Drop-in source is available from our CDN, that you can include in your project as a script tag:
+
+```html
+<script src="https://js.braintreegateway.com/web/dropin/@VERSION/js/dropin.min.js"></script>
````
Should we include:

```html
<script>
braintree.dropin.create({ /* options */ }, callback);
</script>
```

lilaconlee replied:
I interpreted this section as "how to include drop-in in your code" but I could go either way.

crookedneighbor commented on jsdoc/Home.md:
````diff
@@ -0,0 +1,90 @@
+To translate Drop-in into different languages, pass a [supported locale code](/docs/current/module-braintree-web-drop-in.html#.create) in your `create` call:
+
+```js
+braintree.dropin.create({
````
Again, this is different on `npm`

crookedneighbor commented on src/dropin.js:
```diff
@@ -19,6 +19,86 @@ var svgHTML = fs.readFileSync(__dirname + '/html/svgs.html', 'utf8');
+ * <caption>Dynamically enable or disable your submit button based on whether or not the payment method is requestable</caption>
+ * var submitButton = document.querySelector('#submit-button');
+ *
+ * braintree.dropin.create({
```
Different on npm

crookedneighbor commented on src/index.js:
```diff
@@ -9,6 +12,115 @@ var DropinError = require('./lib/dropin-error');
+/**
+ * @static
+ * @function create
+ * @description This function is the entry point for `braintree.dropin`. It is used for creating {@link Dropin} instances.
```
Again, different on `npm`

crookedneighbor commented on src/index.js:
```diff
@@ -9,6 +12,115 @@ var DropinError = require('./lib/dropin-error');
+ *     <script>
+ *       var submitButton = document.querySelector('#submit-button');
+ *
+ *       braintree.dropin.create({
```
`npm`

crookedneighbor commented on src/index.js:
```diff
@@ -9,6 +12,115 @@ var DropinError = require('./lib/dropin-error');
+ *
+ * @example
+ * <caption>Setting up a Drop-in instance to accept credit cards, PayPal, and PayPal Credit</caption>
+ * braintree.dropin.create({
```
`npm`

crookedneighbor commented on src/index.js:
```diff
@@ -9,6 +12,115 @@ var DropinError = require('./lib/dropin-error');
+ * @param {object} [options.paypal] The configuration options for PayPal. To include a PayPal option in your Drop-in integration, include the `paypal` parameter and [enable PayPal in the Braintree Control Panel](https://developers.braintreepayments.com/guides/paypal/testing-go-live/#go-live). To test in Sandbox, you will need to [link a PayPal sandbox test account to your Braintree sandbox accoun ...
+ *
+ * Some of the PayPal configuration options are listed here, but for a full list see the [PayPal Checkout client reference options](http://braintree.github.io/braintree-web/{@pkg bt-web-version}/PayPalCheckout.html#createPayment).
+ * @param {string} [options.paypal.flow] Either `checkout` for a one-time [Checkout with PayPal](https://developers.braintreepayments.com/guides/paypal/checkout-with-paypal/javascript/v3) flow or `vault` for a [Vault flow](https://developers.braintreepayments.com/guides/paypal/vault/javascript/v3). Required when using PayPal.
```
This param is not optional.

lilaconlee replied:
Marked it as optional because the whole `paypal` param is optional, not sure what we usually do there.

crookedneighbor replied:
My opinion for this kind of stuff is that the parent param is options `options.paypal`, but if you include `options.paypal` the `flow` param is _not_ optional. So the flow param shows as required, even if the parent object is optional.

crookedneighbor commented on src/index.js:
```diff
@@ -9,6 +12,115 @@ var DropinError = require('./lib/dropin-error');
+ * @param {function} callback The second argument, `data`, is the {@link Dropin} instance.
+ * @returns {void}
+ * @example
+ * <caption>A full example of accepting credit cards</caption>
```
Should we include a form example as well?

Multiple times this week, I helped people troubleshoot their Drop-in integration where they had their drop-in in a form and it would submit even though they had a listener on the submit button. This is because although we can prevent default on the submit button, it does not prevent default on the form submission.

lilaconlee replied:
Oh definitely great idea.

crookedneighbor commented on src/dropin.js:
```diff
@@ -19,6 +19,86 @@ var svgHTML = fs.readFileSync(__dirname + '/html/svgs.html', 'utf8');
+ * @function
+ * @param {string} event The name of the event to which you are subscribing.
+ * @param {function} handler A callback to handle the event.
+ * @description Subscribes a handler function to a named event. `event` should be {@link HostedFields#event:paymentMethodRequestable|paymentMethodRequestable} or {@link HostedFields#event:noPaymentMethodRequestable|noPaymentMethodRequestable}.
```
Should we put these event names in backticks?

crookedneighbor commented on src/dropin.js:
```diff
@@ -19,6 +19,86 @@ var svgHTML = fs.readFileSync(__dirname + '/html/svgs.html', 'utf8');
+/**
+ * @name Dropin#on
+ * @function
+ * @param {string} event The name of the event to which you are subscribing.
```
There's an enum property in JSDoc I think we can use.

lilaconlee replied:
I just did this the same way as we did in Hosted Fields, down to update it in both places though.

crookedneighbor commented on src/dropin.js:
```diff
@@ -169,6 +249,12 @@ Dropin.prototype._disableErroredPaymentMethods = function () {
+/**
+ * Requests a payment method object, that includes the payment method nonce used by by the [Braintree Server SDKs](https://developers.braintreepayments.com/start/hello-server/). The structure of this object varies by type: a {@link Dropin~cardPaymentMethodPayload|cardPaymentMethodPayload} is returned when the payment method is a card, a {@link Dropin~paypalPaymentMethodPayload|paypalPaymentMethod ...
+ * @public
+ * @param {callback} callback The first argument will be an error if no payment method is available and will otherwise be null. The second argument will be an object containing a payment method nonce; either a {@link Dropin~paypalPaymentMethodPayload|paypalPaymentMethodPayload} or a {@link Dropin~paypalPaymentMethodPayload|paypalPaymentMethodPayload}.
```
I misread this at first as the payment method nonce being a paypal or card payload. Not sure if this needs to be adjusted or not.

crookedneighbor commented on src/dropin.js:
```diff
@@ -227,6 +313,12 @@ Dropin.prototype._getVaultedPaymentMethods = function (callback) {
 };
 
+/**
+ * Cleanly remove anything set up by {@link module:braintree-web-drop-in|dropin.create}. This maybe be useful in a single-page app.
```
maybe -> may be

crookedneighbor commented on src/dropin.js:
```diff
@@ -19,6 +19,86 @@ var svgHTML = fs.readFileSync(__dirname + '/html/svgs.html', 'utf8');
+/**
+ * @typedef {object} Dropin~paymentMethodRequestablePayload
+ * @description The event payload sent from {@link Dropin#on|on} with the {@link Dropin#event:paymentMethodRequestable|paymentMethodRequestable} event.
+ * @property {string} type The type of payment method that is requestable. Either `CreditCard` or `PayPalAccount`.
```
We should be able to use enums here.

lilaconlee commented:
@crookedneighbor addressed most of your comments. Only mentioned the npm naming stuff in the setup section, but I figure we can add more if folks are confused. Didn't add the enums for JSDoc events because I wasn't quite sure what to do there.

crookedneighbor commented:
I thought there was an easy way to add a property to the param stating the possible values, but I can't find it now, so I'm okay leaving it off.

intelliot commented on src/dropin.js:
```diff
@@ -19,6 +19,86 @@ var svgHTML = fs.readFileSync(__dirname + '/html/svgs.html', 'utf8');
+ * @typedef {object} Dropin~cardPaymentMethodPayload
+ * @property {string} nonce The payment method nonce.
+ * @property {object} details Additional account details.
+ * @property {string} details.cardType Type of card, ex: Visa, MasterCard.
```
I would consider using `e.g.` since that's what's used [elsewhere](https://github.com/braintree/braintree-web-drop-in/pull/167/files#diff-1fdf421c05c1140f6d71444ea2b27638R21)

intelliot commented on jsdoc/index.html:
```diff
@@ -0,0 +1,7 @@
+<!doctype html>
+<html>
+  <head>
+    <meta http-equiv="refresh" content="0; url=current">
```
Why do we have this meta tag?

lilaconlee replied:
This redirects to `braintree-web-drop-in/docs/` to `braintree-web-drop-in/docs/current`. It's the same thing we do for `braintree-web` docs.
